idaes/core/util/tables.py

```diff
@@ -89,24 +88,26 @@ def _stream_dict_add(sb, n, i=None):
             key = "{}[{}]".format(n, i)
         stream_dict[key] = sb
 
-    for n in streams.keys():
-        try:
-            if isinstance(streams[n], Arc):
-                for i, a in streams[n].items():
-                    sb = _get_state_from_port(a.ports[1], time_point)
-                    _stream_dict_add(sb, n, i)
-            elif isinstance(streams[n], Port):
-                sb = _get_state_from_port(streams[n], time_point)
-                _stream_dict_add(sb, n)
-            else:
+    for n in streams.keys():        
+        if isinstance(streams[n], Arc):
+            for i, a in streams[n].items():
+                sb = _get_state_from_port(a.ports[1], time_point)
+                _stream_dict_add(sb, n, i)
+        elif isinstance(streams[n], Port):
+            sb = _get_state_from_port(streams[n], time_point)
+            _stream_dict_add(sb, n)
+        else:
+            # _IndexedStateBlock is a private class, so cannot directly test
+            # whether  streams[n] is one or not.
+            try:
                 sb = streams[n][time_point]
-                _stream_dict_add(sb, n)
-        except (AttributeError, KeyError):
-            raise TypeError(
-                f"Unrecognised component type for stream argument {streams[n]}."
-                f" The stream_states_dict function only supports Arcs, "
-                f"Ports or StateBlocks."
-            )
+            except (AttributeError, KeyError) as err:
```

Comment:
Is AttributeError still needed here? At this point I think only a KeyError could be raised.

Reply:
I'm not sure why it was there to begin with. Removing it doesn't cause any tests to fail, which probably means we can get rid of it. 

Unfortunately, pulling on this thread caused some other parts of the code to unravel. Presently, the exception message indicates that the problem is that a StateBlock wasn't passed, when the real error is that it isn't able to be indexed by `time_point`. As a result, this `except` block handles errors for which the real problem was that the user passed in some `time_point` outside of `flowsheet.time`, but doesn't prevent other IndexedBlocks to pass through even though they are the wrong type.  Our options appear to be:

1. Add in a subfunction that crawls up the Pyomo object hierarchy to find the flowsheet, then test `time_point in flowsheet.time`, and make sure it's called in every function that accepts `time_point` as an option. Then explicitly add in testing that `isinstance(sb,StateBlock)` as part of each case.
2. Remove the `except` block and make no pretense that we're protecting the user from themselves.

At this point, I'd rather go with option (2).

Reply:
I would go with 3:

Raise an exception that says the component was either unindexed or the index was not a member of the indexing set. Then let the user work out the problem.